pkg/table/column.go

```diff
@@ -536,6 +537,23 @@ func GetColOriginDefaultValueWithoutStrictSQLMode(ctx expression.BuildContext, c
 	})
 }
 
+// CheckNoDefaultValueForInsert checks if the column has no default value before insert data.
+// CheckNoDefaultValueForInsert extrats the check logic from getColDefaultValueFromNil,
+// since getColDefaultValueFromNil function is public path and both read/write and other places use it.
+// But CheckNoDefaultValueForInsert logic should only check before insert.
+func CheckNoDefaultValueForInsert(sc *stmtctx.StatementContext, col *model.ColumnInfo) error {
+	if mysql.HasNoDefaultValueFlag(col.GetFlag()) && !col.DefaultIsExpr && col.GetDefaultValue() == nil && col.GetType() != mysql.TypeEnum {
```

Comment:
Do we need to check "col.GetOriginDefaultValue() == nil"? May be we can add a test for "alter table ... add column".

Reply:
I'm not sure about this, since this function logic is extracted from getColDefaultValueFromNil, and the original check don't check `col.GetOriginDefaultValue() == nil`.

Reply:
Already add a test for it.

Reply:
`&& col.GetType() != mysql.TypeEnum`
enum type seems special, why?

Reply:
Get it!
 https://github.com/pingcap/tidb/pull/51309/files#diff-5bd22e23c06e1720146774f77761fe2f910d693465e39ac954c94e5f5f3bf4c6R649